utility/resource_loader_compat.cpp

```diff
@@ -0,0 +1,2084 @@
+		}
+
+		Resource *r = Object::cast_to<Resource>(obj);
+		if (!r) {
+			String obj_class = obj->get_class();
+			error = ERR_FILE_CORRUPT;
+			memdelete(obj); //bye
+			ERR_FAIL_V_MSG(RES(), local_path + ":Resource type in resource field not a resource, type is: " + obj_class + ".");
+		}
+
+		if (path != String() && cache_mode != ResourceFormatLoader::CACHE_MODE_IGNORE) {
+			r->set_path(path, cache_mode == ResourceFormatLoader::CACHE_MODE_REPLACE); //if got here because the resource with same path has different type, replace it
+		}
+		r->set_subindex(subindex);
+		res = RES(r);
+	}
+	return res;
+}
+
+RES ResourceLoaderBinaryCompat::get_internal_resource(const int subindex) {
+	for (auto R = internal_index_cache.front(); R; R = R->next()) {
+		if (R->value()->get_subindex() == subindex) {
+			return R->value();
+		}
+	}
+	return RES();
+}
+
+RES ResourceLoaderBinaryCompat::get_internal_resource(const String &path) {
+	if (has_internal_resource(path)) {
+		return internal_index_cache[path];
+	}
+	return RES();
+}
+
+String ResourceLoaderBinaryCompat::get_internal_resource_type(const String &path) {
+	if (has_internal_resource(path)) {
+		return internal_type_cache[path];
+	}
+	return "None";
+}
+
+List<ResourceProperty> ResourceLoaderBinaryCompat::get_internal_resource_properties(const String &path) {
+	if (has_internal_resource(path)) {
+		return internal_index_cached_properties[path];
+	}
+	return List<ResourceProperty>();
+}
+
+bool ResourceLoaderBinaryCompat::has_internal_resource(const String &path) {
+	return internal_index_cache.has(path);
+}
+
+String ResourceLoaderBinaryCompat::get_resource_path(const RES &res) {
+	if (res.is_null()) {
+		return "";
+	}
+	if (res->is_class("FakeResource")) {
+		return ((Ref<FakeResource>)res)->get_real_path();
+	} else if (res->is_class("FakeScript")) {
+		return ((Ref<FakeScript>)res)->get_real_path();
+	} else {
+		return res->get_path();
+	}
+}
+
+bool check_old_type_name(Variant::Type type, int ver_major, String &old_name) {
+	if (4 == VERSION_MAJOR) {
+		return true;
+	}
+	String new_name = Variant::get_type_name(type);
+	if (new_name.begins_with("Packed")) {
+		old_name = new_name.replace_first("Packed", "");
+		return false;
+	}
+	if (ver_major == 2) {
+		if (new_name == "Transform2D") {
+			old_name = "Matrix32";
+			return false;
+		}
+		if (new_name == "Basis") {
+			old_name = "Matrix3";
+			return false;
+		}
+	}
+	return true;
+}
+
+void print_class_prop_list(const StringName &cltype) {
+	List<PropertyInfo> *listpinfo = memnew(List<PropertyInfo>);
+
+	ClassDB::get_property_list(StringName(cltype), listpinfo);
+	if (!listpinfo || listpinfo->size() == 0) {
+		if (listpinfo) {
+			memdelete(listpinfo);
+		}
+		ERR_FAIL_COND_MSG((!listpinfo || listpinfo->size() == 0), "No properties in class of type " + cltype + " ?");
+	}
+	print_line("List of valid properties in class" + cltype + " : \n[");
+	for (auto I = listpinfo->front(); I; I->next()) {
```

Comment:
```suggestion
	for (auto I = listpinfo->front(); I; I = I->next()) {
```
Infinite loop, iterator is not update.